compiler/aarch64/codegen/OMRCodeGenerator.cpp

```diff
@@ -19,52 +19,313 @@
    }
 
 void
 OMR::ARM64::CodeGenerator::endInstructionSelection()
    {
-   TR_ASSERT(false, "Not implemented yet.");
+   if (_returnTypeInfoInstruction != NULL)
+      {
+      _returnTypeInfoInstruction->setSourceImmediate(static_cast<uint32_t>(self()->comp()->getReturnInfo()));
+      }
    }
 
 void
 OMR::ARM64::CodeGenerator::doRegisterAssignment(TR_RegisterKinds kindsToAssign)
    {
-   TR_ASSERT(false, "Not implemented yet.");
+   // Registers are assigned in backward direction
+
+   TR::Compilation *comp = self()->comp();
+
+   TR::Instruction *instructionCursor = self()->getAppendInstruction();
+
+   if (!comp->getOption(TR_DisableOOL))
+      {
+      TR::list<TR::Register*> *spilledRegisterList = new (self()->trHeapMemory()) TR::list<TR::Register*>(getTypedAllocator<TR::Register*>(comp->allocator()));
+      self()->setSpilledRegisterList(spilledRegisterList);
+      }
+
+   if (self()->getDebug())
+      self()->getDebug()->startTracingRegisterAssignment();
+
+   while (instructionCursor)
+      {
+      TR::Instruction *prevInstruction = instructionCursor->getPrev();
+
+      self()->tracePreRAInstruction(instructionCursor);
+
+      self()->setCurrentBlockIndex(instructionCursor->getBlockIndex()); // ppc
+
+      instructionCursor->assignRegisters(TR_GPR);
+
+      // Maintain Internal Control Flow Depth
+      // Track internal control flow on labels
+      if (instructionCursor->isLabel())
+         {
+         TR::ARM64LabelInstruction *li = (TR::ARM64LabelInstruction *)instructionCursor;
+
+         if (li->getLabelSymbol() != NULL)
+            {
+            if (li->getLabelSymbol()->isStartInternalControlFlow())
+               {
+               self()->decInternalControlFlowNestingDepth();
+               }
+            if (li->getLabelSymbol()->isEndInternalControlFlow())
+               {
+               self()->incInternalControlFlowNestingDepth();
+               }
+            }
+         }
+
+      self()->freeUnlatchedRegisters();
+      self()->buildGCMapsForInstructionAndSnippet(instructionCursor);
+
+      self()->tracePostRAInstruction(instructionCursor);
+
+      instructionCursor = prevInstruction;
+      }
+
+   if (self()->getDebug())
+      {
+      self()->getDebug()->stopTracingRegisterAssignment();
+      }
+   }
+
+static int32_t identifyFarConditionalBranches(int32_t estimate, TR::CodeGenerator *cg)
+   {
+   TR_Array<TR::ARM64ConditionalBranchInstruction *> candidateBranches(cg->trMemory(), 256);
+   TR::Instruction *cursorInstruction = cg->getFirstInstruction();
+
+   while (cursorInstruction)
+      {
+      TR::ARM64ConditionalBranchInstruction *branch = cursorInstruction->getARM64ConditionalBranchInstruction();
+      if (branch != NULL)
+         {
+         if (abs(branch->getEstimatedBinaryLocation() - branch->getLabelSymbol()->getEstimatedCodeLocation()) > 512 * 1024)
```

Comment:
The code came from OMRCodeGenerator.cpp of p and arm32 versions.  Their branch instructions can reach +/-32KB address range.
The first loop puts branch instructions with the displacement larger than +/-16KB (half of 32KB) to candidate list.  The second loop sets the FarRelocation flag by checking with +/-32KB range.

AArch64 conditional branch can reach +/-1MB range, and I used 512KB (half of 1MB) instead of 16KB for p/arm32.